pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,296 @@
+         * @return The delta time transcurred since the {@link TimerEntry} began.
+         */
+        /* package */ long accumulate(final TimerEntry timerEntry, final long extraData) {
+            final long delta = System.nanoTime() - timerEntry.start;
+            
+            totalTime.getAndAdd(delta);
+            selfTime.getAndAdd(delta - timerEntry.inNestedOperations);
+            callCount.getAndIncrement();
+            extraDataCounter.getAndAdd(extraData);
+            
+            return delta;
+        }
+        
+        /**
+         * Merges the times (and only the times) from another {@link TimedResult} into self.
+         * @param timedResult The {@link TimedResult} to merge
+         */
+        /* package */ void mergeTimes(final TimedResult timedResult) {
+            totalTime.getAndAdd(timedResult.totalTime.get());
+            selfTime.getAndAdd(timedResult.selfTime.get());
+        }
+    }
+    
+    /**
+     * A unique identifier for a timed operation
+     */
+    /* package */ static class TimedOperationKey {
+        /* package */ final TimedOperationCategory category;
+        /* package */ final String label;
+        
+        /* package */ TimedOperationKey(final TimedOperationCategory category, final String label) {
+            this.category = category;
+            this.label = label;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + ((category == null) ? 0 : category.hashCode());
+            result = prime * result + ((label == null) ? 0 : label.hashCode());
+            return result;
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            TimedOperationKey other = (TimedOperationKey) obj;
+            if (category != other.category) {
+                return false;
+            }
+            return Objects.equals(label, other.label);
+        }
+
+        @Override
+        public String toString() {
+            return "TimedOperation [category=" + category + ", label=" + label + "]";
+        }
+    }
+    
+    /**
+     * A standard timed operation implementation.
+     */
+    private static class TimedOperationImpl implements TimedOperation {
+
+        private boolean closed = false;
+
+        @Override
+        public void close() {
+            stop(0);
+        }
+
+        @Override
+        public void stop(int extraDataCounter) {
+            if (closed) {
+                return;
+            }
+
+            closed = true;
+            TimeTracker.finishOperation(extraDataCounter);
+        }
+
+        @Override
+        public void stop() {
+            stop(0);
+        }
```

Comment:
These `stop` method semantics are essentially just `close` with another name.  And `stop()` is utterly redundant.  Maybe keep one and rename to `close(int)`.  The docs could reflect behavior upon repeated calls.

The semantics are different that what I was originally thinking, but it works for current usage.  It's ok, maybe just tighten it up a little.